### PR TITLE
Improve make() capacity allocations

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR reviews the usage of make() within prysm and sets the capcity or length to the appropriate size when known.

This should reduce the total number of allocations at runtime for Prysm.

**Which issues(s) does this PR fix?**

N/A - memory usage

**Other notes for review**

Includes #6508, please review and merge that first before merging this.

### CODE REVIEW

File: beacon-chain/operations/slashings/service.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"sort"
 
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go:9]: extra space

File: beacon-chain/operations/slashings/service.go [hunk 2]
```diff
@@ -36,7 +38,9 @@ func (p *Pool) PendingAttesterSlashings(ctx context.Context, state *beaconstate.
 	numPendingAttesterSlashings.Set(float64(len(p.pendingAttesterSlashing)))
 
 	included := make(map[uint64]bool)
-	pending := make([]*ethpb.AttesterSlashing, 0, params.BeaconConfig().MaxAttesterSlashings)
+	// Allocate pending slice with a capacity of min(len(p.pendingAttesterSlashing), maxAttesterSlashings)
+	// since the array cannot exceed the max and is typically less than the max value.
+	pending := make([]*ethpb.AttesterSlashing, 0, mathutil.Min(uint64(len(p.pendingAttesterSlashing)), params.BeaconConfig().MaxAttesterSlashings))
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go:43]: Shouldn't this be max? it cant go beyond `MaxAttesterSlashings` right?

File: beacon-chain/operations/voluntaryexits/service.go [hunk 3]
```diff
@@ -5,6 +5,8 @@ import (
 	"sort"
 	"sync"
 
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+
```
Comment (terencechain) [inline beacon-chain/operations/voluntaryexits/service.go:9]: extra space

File: beacon-chain/operations/slashings/service.go [hunk 4]
```diff
@@ -75,7 +79,9 @@ func (p *Pool) PendingProposerSlashings(ctx context.Context, state *beaconstate.
 	// Update prom metric.
 	numPendingProposerSlashings.Set(float64(len(p.pendingProposerSlashing)))
 
-	pending := make([]*ethpb.ProposerSlashing, 0, params.BeaconConfig().MaxProposerSlashings)
+	// Allocate pending slice with a capacity of min(len(p.pendingProposerSlashing), maxProposerSlashings)
+	// since the array cannot exceed the max and is typically less than the max value.
+	pending := make([]*ethpb.ProposerSlashing, 0, mathutil.Min(uint64(len(p.pendingProposerSlashing)), params.BeaconConfig().MaxProposerSlashings))
```
Comment (terencechain) [inline beacon-chain/operations/slashings/service.go:84]: Same here. Max?

File: beacon-chain/operations/voluntaryexits/service.go [hunk 5]
```diff
@@ -34,7 +36,10 @@ func NewPool() *Pool {
 func (p *Pool) PendingExits(state *beaconstate.BeaconState, slot uint64) []*ethpb.SignedVoluntaryExit {
 	p.lock.RLock()
 	defer p.lock.RUnlock()
-	pending := make([]*ethpb.SignedVoluntaryExit, 0)
+
+	// Allocate pending slice with a capacity of min(len(p.pending), maxVoluntaryExits) since the
+	// array cannot exceed the max and is typically less than the max value.
+	pending := make([]*ethpb.SignedVoluntaryExit, 0, mathutil.Max(uint64(len(p.pending)), params.BeaconConfig().MaxVoluntaryExits))
```
Comment (terencechain) [inline beacon-chain/operations/voluntaryexits/service.go:42]: This is right. It uses max

File: beacon-chain/operations/attestations/kv/block.go [hunk 6]
```diff
@@ -48,10 +48,10 @@ func (p *AttCaches) SaveBlockAttestations(atts []*ethpb.Attestation) error {
 
 // BlockAttestations returns the block attestations in cache.
 func (p *AttCaches) BlockAttestations() []*ethpb.Attestation {
-	atts := make([]*ethpb.Attestation, 0)
-
 	p.blockAttLock.RLock()
 	defer p.blockAttLock.RUnlock()
+
+	atts := make([]*ethpb.Attestation, 0, len(p.blockAtt))
```
Comment (0xKiwi) [inline beacon-chain/operations/attestations/kv/block.go:54]: This doesn't look safe, looks like p.blockAtt is a double array?

File: beacon-chain/operations/slashings/service.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:43]: That’s true, but there’s no point in allocating extra data if there aren’t no slashing.

File: beacon-chain/operations/slashings/service.go (hunk 4 above)
Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:84]: Same comment. No need to allocate extra data when there is less than max slashings available. 

Is there anything I can make more clear in the comment above?

Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:84]: Same comment. No need to allocate extra data when there is less than max slashings available. 

Is there anything I can make more clear in the comment above?

Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:84]: Same comment. No need to allocate extra data when there is less than max slashings available. 

Is there anything I can make more clear in the comment above?

File: beacon-chain/operations/voluntaryexits/service.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/operations/voluntaryexits/service.go:42]: This is wrong, it should be min. I’ll fix

File: beacon-chain/operations/slashings/service.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/operations/slashings/service.go:43]: You are right. For some reason my brain was too stubborn to think that...

File: beacon-chain/operations/slashings/service.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/operations/slashings/service.go:84]: This is good

File: beacon-chain/operations/attestations/kv/block.go (hunk 6 above)
Comment (nisdas) [reply beacon-chain/operations/attestations/kv/block.go:54]: which part isnt safe ? this just adds in the capacity of the slice

File: beacon-chain/operations/slashings/service.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:43]: It's ok. I did the same thing somewhere when I wrote Max lol

File: beacon-chain/operations/slashings/service.go (hunk 4 above)
Comment (prestonvanloon) [reply beacon-chain/operations/slashings/service.go:84]: Sorry for spam, i was on mobile! lol

File: beacon-chain/operations/voluntaryexits/service.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/operations/voluntaryexits/service.go]: Done!

File: beacon-chain/operations/attestations/kv/block.go (hunk 6 above)
Comment (prestonvanloon) [reply beacon-chain/operations/attestations/kv/block.go:54]: Oh shoot, I think @0xKiwi has a good point. This p.blockAtt is a map of arrays. Let me double check this is correct...

Comment (prestonvanloon) [reply beacon-chain/operations/attestations/kv/block.go:54]: Yeah.. I'm going to revert this one. It's not unsafe but it's also not correct at all.

We might have a map like 
0x01 -> [att, att, att]
0x02 -> [att, att, att]
0xAB -> [att, att, att]

The code here would create and only allocate enough memory for 3 attestations when the true number was 9.

File: beacon-chain/operations/attestations/kv/aggregated.go [hunk 7]
```diff
@@ -111,10 +111,10 @@ func (p *AttCaches) SaveAggregatedAttestations(atts []*ethpb.Attestation) error
 
 // AggregatedAttestations returns the aggregated attestations in cache.
 func (p *AttCaches) AggregatedAttestations() []*ethpb.Attestation {
-	atts := make([]*ethpb.Attestation, 0)
-
 	p.aggregatedAttLock.RLock()
 	defer p.aggregatedAttLock.RUnlock()
+
+	atts := make([]*ethpb.Attestation, 0, len(p.aggregatedAtt))
```
Comment (prestonvanloon) [inline beacon-chain/operations/attestations/kv/aggregated.go:117]: Same feedback here from @0xKiwi. I'll revert this one as well and revisit later (maybe).

**Pull request merged**